rfcs/0000-state-separation.md

```diff
@@ -0,0 +1,706 @@
+    State::Example,
+    Example {
+        scene: self.prefab.as_ref().unwrap().clone(),
+    },
+);
+```
+
+## Refactor amethyst_test
+
+Since a large number of things changed in how applications are constructed, `amethyst_test` needed to be refactored as well.
+
+I'm not gonna cover too much about what happened, but if you are curious, the tests and API docs should have been updated.
+
+But briefly:
+* All methods prefixed with `do_*` are run _in order_.
+* `with_effect`, `with_assertion`, and `with_setup` all were the same alias. So they have been replaced with `do_fn`.
+* You can switch state using `do_state(&S)`.
+* A `do_wait(usize)` function has been introduced to wait a number of frames before progressing.
+
+A simple test case would now look like this:
+
+```rust
+AmethystApplication::blank()
+    .with_state("first", FunctionState::new(|world| {
+        world.add_resource(ApplicationResource);
+    }))
+    .do_state("first")
+    .do_wait(1)
+    .do_fn(|world| {
+        /// Reading the resource should be fine.
+        world.read_resource::<ApplicationResource>();
+    })
+    .run()
+    .expect("an error")
+```
+
+## Example Application
+
+This is a trimmed version of the `renderable` example, using the proposed API changes:
+
+```rust
+#[macro_use]
+extern crate amethyst;
+
+use amethyst::{ /* snip */ };
+
+type MyPrefabData = BasicScenePrefab<Vec<PosNormTex>>;
+
+#[derive(State, Debug, Clone)]
+enum State {
+    Loading,
+    Example,
+}
+
+#[derive(Default)]
+struct Loading {
+    progress: ProgressCounter,
+    prefab: Option<Handle<Prefab<MyPrefabData>>>,
+}
+
+struct Example {
+    scene: Handle<Prefab<MyPrefabData>>,
+}
+
+impl StateHandler<State, StateEvent> for Loading {
+    fn on_start(&mut self, world: &mut World) {
+        self.prefab = Some(world.exec(|loader: PrefabLoader<MyPrefabData>| {
+            loader.load("prefab/renderable.ron", RonFormat, (), &mut self.progress)
+        }));
+
+        world.exec(|mut creator: UiCreator| {
+            creator.create("ui/fps.ron", &mut self.progress);
+            creator.create("ui/loading.ron", &mut self.progress);
+        });
+    }
+
+    fn update(&mut self, world: &mut World) -> Trans<State> {
+        match self.progress.complete() {
+            Completion::Failed => {
+                println!("Failed loading assets: {:?}", self.progress.errors());
+                Trans::Quit
+            }
+            Completion::Complete => {
+                println!("Assets loaded, swapping state");
+
+                if let Some(entity) = world.exec(|finder: UiFinder| finder.find("loading")) {
+                    let _ = world.delete_entity(entity);
+                }
+
+                world.write_resource::<States<_, _>>().new_state(
+                    State::Example,
+                    Example {
+                        scene: self.prefab.as_ref().unwrap().clone(),
+                    },
+                );
+
+                Trans::Push(State::Example)
```

Comment:
It feels a tad silly to specify `State::Example` twice here, though I'm not 100% certain what we should do instead.  Perhaps this is a demonstration that our StateMachine needs to be rethought a bit.

Reply:
Maybe. `new_state` could in theory return `Trans::Push` with the state that was just written. But this is supposed to be less common than other uses.

Reply:
That's kind of where I feel this RFC makes a few blunders.  All of our examples are really simple and make minimal use of the state machine, but the state machine is designed to accommodate full scale games.

It's kind of sad that the renderable example was the only one that gave you any grief on this to be honest.  That alone serves as a great demonstration that our examples aren't comprehensive enough.

State machine is mostly there to help users navigate through UI and various game modes.  You could imagine it kind of like this for a full scale game

Menu -> Multiplayer Menu -> Multiplayer lobby -> Play Game

Where each word is a state.  After the game is finished we'd pop the Play Game state, returning to the lobby in this example.  Then when we're done playing with this group we'd return to the multiplayer menu, so on so forth.

This is a super common need for full scale games, and isn't demonstrated at all in any of our examples.

Reply:
That's fair, that was written mostly in response to "less common than other uses"  I would prefer it if `new_state` returned the `Trans::push` as you said earlier.  Sorry for being a bit hard on the RFC in my opening statement, that wasn't fair to you or the RFC.